### ISSUE TITLE
Better PunktTrainer

### ISSUE THREAD

**Issue by alvations** [nltk/nltk#2008]
While trying to retrain a sentence tokenizer model with `PunktTokenizer`, the NLTK code took up >200GB of RAM and a lot of swap and doesn't seem to end after 2 days of training. 

```python
import pickle
from nltk.tokenize.punkt import PunktSentenceTokenizer

tokenizer = PunktSentenceTokenizer()

with open('wikipedia-en.txt') as fin:
    text = fin.read()

tokenizer.train(text)

with open('wiki-en.pkl') as fout:
    pickle.dump(tokenizer, fout)
```

The code seem inefficient for the massive data set we have now to train better Punkt models. 

Let's keep this issue to track Punkt related issue until we get a better version of the algorithm.

**Comment by alvations** [comment]
Replace `_pair_iter` with `bigrams`
====

This seems to be like the ngram function we have in `nltk.util.ngrams`: https://github.com/nltk/nltk/blob/develop/nltk/tokenize/punkt.py#L306 

We shouldn't be maintaining both `_pair_iter(it)` and `bigrams(it)`.

Replace `counter` with `bounter`
====

Our FreqDist object based on `collections.Counter` has a high memory footprint, we should replace it with [`bounter`](https://github.com/RaRe-Technologies/bounter).

There're several instances of FreqDist usage in the Punkt code and the `self._collocation_fdist` should benefit most from `bounter` with a lower memory needed, i.e. https://github.com/nltk/nltk/blob/develop/nltk/tokenize/punkt.py#L631 

Constants (Magic numbers) should be parameters
====

There are several instances where constant values in `PunktTrainer` are used to set thresholds / backoff for the algorithm, they are not exposed to the users and set as fixed constants. They should be tune-able parameters. E.g. https://github.com/nltk/nltk/blob/develop/nltk/tokenize/punkt.py#L671 


Monitoring and optimizing for-loops
====

The for-loops in [`_train_tokens()`](https://github.com/nltk/nltk/blob/develop/nltk/tokenize/punkt.py#L739) iterates through the dataset in multi-pass. 

Most probably the multi-pass cannot be avoided but we should be able to optimize the passes with some parallelization and if possibly don't count but approximate the counts with some string based algorithms (we've got to look into this). 

There are some for-loops that looks like they are independent of each other, these could be parallelized too. 

Clean up Py2 and Py3 string compatibility code
====

The `nltk.token.punkt.py` contains a variety of checks for text types and compatibility checks, they should unify to either using the native `nltk.compat` code or `six`:

```
from six import string_types
from nltk.compat import unicode_repr, python_2_unicode_compatible
```

**Comment by jnothman** [comment]
You're missing the fact that the `PunktTrainer` has a facility to manage memory usage due to infrequent elements. It's a method called `freq_threshold` that is intended for calling between batches of training. Yes, it should be more clearly documented.

We have had punkt models trained on Wikipedia for years, but when I first did so, it was using an occasional `freq_threshold` to prune useless statistics.

`bounters` should benefit hugely, in a similar vein, but I think your other proposed changes are largely cosmetic. Parameters for things like `ABBREV` would add noise to the interface when a user can override a constant by merely subclassing.

**Comment by alvations** [comment]
Yes, I was reading the code to understand what's happening while listing things that can be improved. But `bounter` replacement is the main goal here =)

I find the Punkt word tokenizer useful too, maybe exposing it out would be good instead of hiding it inside the `PunktLanguageVars`.

Cool, I didn't know that there's a `freq_threshold` parameter! Still reading the code.

**Comment by jnothman** [comment]
I think it was previously more exposed and caused confusion...

**Comment by alvations** [comment]
It does ~50% of the cleaning task that I've been doing with other regexes for non-English languages though. I think I've to look at the code history too.

Interesting: #746 and 0b91a7160717faa2fe93d42f7c6bba735f6dd48a

Unfortunately, going from `FreqDist` -> `bounter` isn't an easy fix though. First problem encountered, FreqDist allows any object to be the key while bounter only allows unicode object or byte buffer.

**Comment by alvations** [comment]
Collapsing multiply loops might help. In the current code, `PunktTrainer` is looping through to:

1. [Find the frequency of each case-normalized type](https://github.com/nltk/nltk/blob/develop/nltk/tokenize/punkt.py#L739)

2. Add/Remove type-based abbreviations based on the word type's heuristics and thresholds](https://github.com/nltk/nltk/blob/develop/nltk/tokenize/punkt.py#L747)
    - requires the global counts from step 1. 

3. [Creates a first pass annotation to add info about whether a token](https://github.com/nltk/nltk/blob/develop/nltk/tokenize/punkt.py#L765) is a (i) `sentbreak`, (ii) `ellipsis`, (iii)`period_final`
    - requires the updated abbreviation list from step 2

4. [Get the orthography features](https://github.com/nltk/nltk/blob/develop/nltk/tokenize/punkt.py#L765)
    - requires step 3 `token.ellipsis` information from step 3
    - but doesn't need global information so can be computed directly after step 3


5. [Find the sentence breaks](https://github.com/nltk/nltk/blob/develop/nltk/tokenize/punkt.py#L771) 
   - requires `tokens.sentbreak` information from step 3
   - but doesn't need global information so can be computed directly after step 3 and 4


6. [Iterate the pairs of token bigrams and update (i) `token.type_no_period` information, (ii) `_sent_starter_fdist` and (iii) `_collocation_fdist` distributions](https://github.com/nltk/nltk/blob/develop/nltk/tokenize/punkt.py#L774)

**Comment by echan00** [comment]
> You're missing the fact that the PunktTrainer has a facility to manage memory usage due to infrequent elements. It's a method called freq_threshold that is intended for calling between batches of training. Yes, it should be more clearly documented.

@alvations  @jnothman How would I use freq_threshold and batch training?  I have ~60,000 TXT files I want to train with.

This is my current code:

```
### TRAINING
text = ''
for file in glob.glob("*.txt"):
	f = io.open("./"+file,"r",encoding="utf8")
	text += f.read()
	f.close()

trainer = PunktTrainer()
trainer.INCLUDE_ALL_COLLOCS = True
trainer.INCLUDE_ABBREV_COLLOCS = True
trainer.train(text) 
tokenizer = PunktSentenceTokenizer(trainer.get_params())
```

Does this look correct?  How do I know what freq_thresholds to use?

```
text = ''
for file in glob.glob("*.txt"):
	f = io.open("./"+file,"r",encoding="utf8")
	text = f.read()
	trainer = PunktTrainer()
	trainer.INCLUDE_ALL_COLLOCS = True
	trainer.INCLUDE_ABBREV_COLLOCS = True
	trainer.freq_threshold()
        trainer.train(text, finalize=False)
	f.close()

tokenizer = PunktSentenceTokenizer(trainer.get_params())
```
